lib/ProductOpener/Ecoscore.pm

```diff
@@ -0,0 +1,240 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+=head1 NAME
+
+ProductOpener::Ecoscore - compute the Ecoscore environemental grade of a food product
+
+=head1 SYNOPSIS
+
+C<ProductOpener::Ecoscore> is used to compute the Ecoscore environmental grade
+of a food product.
+
+=head1 DESCRIPTION
+
+The modules implements the Eco-Score computation as defined by a collective that Open Food Facts is part of.
+
+It is based on the French AgriBalyse V3 database that contains environmental impact values for 2500 food product categories.
+
+AgriBalyse provides Life Cycle Analysis (LCA) values for food products categories,
+and some adjustments to the score are made for actual specific products using data about labels, origins of ingredients, packagings etc.
+
+=cut
+
+package ProductOpener::Ecoscore;
+
+use utf8;
+use Modern::Perl '2017';
+use Exporter    qw< import >;
+
+use Log::Any qw($log);
+
+BEGIN
+{
+	use vars       qw(@ISA @EXPORT_OK %EXPORT_TAGS);
+	@EXPORT_OK = qw(
+
+		&load_agribalyse_data
+		&compute_ecoscore
+
+		);    # symbols to export on request
+	%EXPORT_TAGS = (all => [@EXPORT_OK]);
+}
+
+use vars @EXPORT_OK ;
+
+use ProductOpener::Config qw/:all/;
+use ProductOpener::Tags qw/:all/;
+
+use Text::CSV();
+
+my %agribalyse = ();
+
+=head1 FUNCTIONS
+
+=head2 load_agribalyse_data( $product_ref )
+
+Loads the AgriBalyse database.
+
+=cut
+
+sub load_agribalyse_data() {
+
+	my $agribalyse_details_by_step_csv_file = $data_root . "/ecoscore/agribalyse/AGRIBALYSE3.0.1_vf.csv.2";
+	
+	my $headers_ref;
+	my $rows_ref = [];
+
+	my $encoding = "UTF-8";
+
+	$log->debug("opening agribalyse CSV file", { file => $agribalyse_details_by_step_csv_file }) if $log->is_debug();
+
+	my $csv_options_ref = { binary => 1, sep_char => "," };    # should set binary attribute.
+
+	my $csv = Text::CSV->new ( $csv_options_ref )
+		or die("Cannot use CSV: " . Text::CSV->error_diag ());
+
+	if (open (my $io, "<:encoding($encoding)", $agribalyse_details_by_step_csv_file)) {
+
+		# @$headers_ref = $csv->header ($io, { detect_bom => 1 });
+		# the header function crashes with some csv files... use getline instead
+		my $row_ref;
+
+		# Skip 3 first lines
+		$csv->getline ($io);
+		$csv->getline ($io);
+		$csv->getline ($io);
+
+		while ($row_ref = $csv->getline ($io)) {
+			$agribalyse{$row_ref->[0]} = {
+				name_fr => $row_ref->[4], # Nom du Produit en Français
+				name_en => $row_ref->[5], # LCI Name
+				dqr => $row_ref->[6], # DQR (data quality rating)
+				# warning: the AGB file has a hidden H column
+				ef_agriculture => $row_ref->[8], # Agriculture
+				ef_processing => $row_ref->[9], # Transformation
+				ef_packaging => $row_ref->[10], # Emballage
+				ef_transportation => $row_ref->[11], # Transport
+				ef_distribution => $row_ref->[12], # Supermarché et distribution
+				ef_consumption => $row_ref->[13], # Consommation
+				ef_total => $row_ref->[14], # Total
```

Comment:
Alternative to the header is [bind_columns](https://metacpan.org/pod/Text::CSV#bind_columns), which according to the [documentation is also faster than using hash refs](https://metacpan.org/pod/Text::CSV#getline_hr). So if we get the header row like this:

```perl
$. = 0;
do { <$io> } until $. == 7 || eof;

my @hdrs = @{$csv->getline($io)};
my $row_ref = {};
$csv->bind_columns(\@{$row_ref}{@hdrs});
```

This can be changed to this:
```suggestion

		while ($csv->getline ($io)) {
			$agribalyse{$row_ref->{'Code AGB'}} = {
				name_fr => $row_ref->{'Nom du Produit en Français'},
				name_en => $row_ref->{'LCI Name' },
				dqr => $row_ref->{'DQR'}, # data quality rating
				# warning: the AGB file has a hidden H column
				ef_agriculture => $row_ref->{'Agriculture'},
				ef_processing => $row_ref->{'Transformation'},
				ef_packaging => $row_ref->{'Emballage'},
				ef_transportation => $row_ref->{'Transport'}, 
				ef_distribution => $row_ref->{'Supermarché et distribution'},
				ef_consumption => $row_ref->{'Consommation'},
				ef_total => $row_ref->{'Total'},
```
Which is more readable if nothing else.

Reply:
I tried it, but unfortunately it does not work because the column names are not unique. There are multiple "Total" columns... :-(

Reply:
Ah. You're right. Looks like it picked the last one total column when I tried it.